From 3a4fba3de8417a361d532b760b5c21b3bce16c11 Mon Sep 17 00:00:00 2001
From: Tony Camuso <tcamuso@redhat.com>
Date: Fri, 8 Aug 2014 18:14:40 -0400
Subject: [char] ipmi: fix timeout calculation when bmc is disconnected

Message-id: <1407521687-4858-7-git-send-email-tcamuso@redhat.com>
Patchwork-id: 87551
O-Subject: [RHEL7.1 PATCH 06/13] ipmi: fix timeout calculation when bmc is disconnected
Bugzilla: 1111214
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1111214

Cherry-picked without conflicts from the following upstream commit.

commit e21404dc0ac7ac971c1e36274b48bb460463f4e5
Author: Xie XiuQi <xiexiuqi@huawei.com>
Date:   Fri Jan 24 14:00:52 2014 -0600

    ipmi: fix timeout calculation when bmc is disconnected

    Loading ipmi_si module while bmc is disconnected, we found the timeout
    is longer than 5 secs.  Actually it takes about 3 mins and 20
    secs.(HZ=250)

    error message as below:
      Dec 12 19:08:59 linux kernel: IPMI BT: timeout in RD_WAIT [ ] 1 retries left
      Dec 12 19:08:59 linux kernel: BT: write 4 bytes seq=0x01 03 18 00 01
      [...]
      Dec 12 19:12:19 linux kernel: IPMI BT: timeout in RD_WAIT [ ]
      Dec 12 19:12:19 linux kernel: failed 2 retries, sending error response
      Dec 12 19:12:19 linux kernel: IPMI: BT reset (takes 5 secs)
      Dec 12 19:12:19 linux kernel: IPMI BT: flag reset [ ]

    Function wait_for_msg_done() use schedule_timeout_uninterruptible(1) to
    sleep 1 tick, so we should subtract jiffies_to_usecs(1) instead of 100
    usecs from timeout.

    Reported-by: Hu Shiyuan <hushiyuan@huawei.com>
    Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
    Signed-off-by: Corey Minyard <cminyard@mvista.com>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: Tony Camuso <tcamuso@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/char/ipmi/ipmi_si_intf.c b/drivers/char/ipmi/ipmi_si_intf.c
index 97f9388..2312a82 100644
--- a/drivers/char/ipmi/ipmi_si_intf.c
+++ b/drivers/char/ipmi/ipmi_si_intf.c
@@ -2709,7 +2709,7 @@ static int wait_for_msg_done(struct smi_info *smi_info)
       smi_result == SI_SM_CALL_WITH_TICK_DELAY) {
    schedule_timeout_uninterruptible(1);
    smi_result = smi_info->handlers->event(
-    smi_info->si_sm, 100);
+    smi_info->si_sm, jiffies_to_usecs(1));
   } else if (smi_result == SI_SM_CALL_WITHOUT_DELAY) {
    smi_result = smi_info->handlers->event(
     smi_info->si_sm, 0);
-- 
1.7.1